### DISCUSSION TITLE
Tunnel init cmd completion response vs IKE SA UP/DOWN response from davici

### DISCUSSION THREAD

**Discussion by hbhokare** [strongswan/strongswan#506] [Q&A]
Hi,

I am trying to do tunnel init (CONN INIT) from my application and waiting for tunnel init (CONN INIT) completion response from davici side. Can we expect the IKE SA UP and immediately after that IKE SA DOWN because of some network issue or other reason when I was waiting for init command completion Or Is IKE SA UP always gives us a guarantee that tunnel init command initiated from the application is got completed? 

Thanks,
-Harish

**Comment by tobiasbrunner** [comment]
Sorry, I don't understand your question.

**Reply by hbhokare** [reply]
Sorry, Let me try it again.
We are using davici code to implement IPSec in our application. We have the following questions about the API completion. 
1. if I issue initiate tunnel API, and it does not complete ( I mean API is in progress state ) Can I expect the IEK SA UP event before I get API resp complete.( I mean before I get API resp as a success)
2. If answer to first question is yes then is there any guarantee from davici that we won't get IKE SA DOWN before the tunnel initiate API complete and after we receive IKE SA UP event.  

Hope I could explain it in better way. 

Thanks in advance!

**Reply by tobiasbrunner** [reply]
You might want to read at my response [here](https://github.com/strongswan/strongswan/discussions/412#discussioncomment-758689) to a similar question regarding termination. It's the same here, that is, the ike-updown event is queued before the command response, which in turn is queued before a potential ike-updown event if the SA goes down shortly afterwards, but connections are notified asynchronously.

**Reply by Thermi** [reply]
You can use the information that an IKE_SA ID is unique (unless it overflows, but that's unlikely to happen before you restart the daemon) to accurately track the status of an IKE_SA, independent of the order you receive the messages in (naturally, when you receive information via explicite initiates or the log, an IKE_SA must first be UP before it can be shown as DOWN via the log).

**Reply by tobiasbrunner** [reply]
Could be tricky because you won't know the unique ID of the SA you initiated beforehand (unless you know the CHILD_SA is created under an existing IKE_SA or you maybe determine the unique ID via log handler). Also note that `initiate()` commands currently don't track "their" CHILD_SA, they are happy with any CHILD_SA getting established under the tracked IKE_SA (i.e. if two vici clients concurrently initiate CHILD_SAs under the same IKE_SA, both return successfully as soon as the first CHILD_SA gets established).

**Reply by hbhokare** [reply]
Thanks Tobias and Thermi!
